logserver/src/main/java/alluxio/logserver/AlluxioLog4jSocketNode.java

```diff
@@ -0,0 +1,143 @@
+    mLogServerProcess = Preconditions.checkNotNull(process,
+        "The log server process cannot be null.");
+    mSocket = Preconditions.checkNotNull(socket, "Client socket cannot be null");
+  }
+
+  @Override
+  public void run() {
+    LoggerRepository hierarchy = null;
+    LoggingEvent event;
+    Logger remoteLogger;
+    try (ObjectInputStream objectInputStream = new ObjectInputStream(
+        new BufferedInputStream(mSocket.getInputStream()))) {
+      while (!Thread.currentThread().isInterrupted()) {
+        try {
+          event = (LoggingEvent) objectInputStream.readObject();
+        } catch (ClassNotFoundException e1) {
+          throw new RuntimeException("Class of serialized object cannot be found.", e1);
+        } catch (IOException e1) {
+          throw new RuntimeException("Cannot read object from stream due to I/O error.", e1);
+        }
+        if (hierarchy == null) {
+          hierarchy = configureHierarchy(
+              event.getMDC(AlluxioRemoteLogFilter.REMOTE_LOG_MDC_APPENDER_NAME_KEY).toString());
+        }
+        remoteLogger = hierarchy.getLogger(event.getLoggerName());
+        if (event.getLevel().isGreaterOrEqual(remoteLogger.getEffectiveLevel())) {
+          remoteLogger.callAppenders(event);
+        }
+      }
+    } catch (IOException e) {
+      // Something went wrong, cannot recover.
+      throw new RuntimeException("Cannot open ObjectInputStream due to I/O error.", e);
+    } finally {
+      try {
+        mSocket.close();
+      } catch (Exception e) {
+        // Log the exception caused by closing socket.
+        LOG.warn("Failed to close client socket.");
+      }
+    }
+  }
+
+  /**
+   * Configure a {@link Hierarchy} instance used to retrive logger by name and maintain the logger
+   * hierarchy. {@link AlluxioLog4jSocketNode} instance can retrieve the logger to log incoming
+   * {@link org.apache.log4j.spi.LoggingEvent}s.
+   *
+   * @param logAppenderName name of the appender to use for this client
+   * @return a {@link Hierarchy} instance to retrieve logger
+   * @throws IOException if fails to create an {@link FileInputStream} to read log4j.properties
+   */
+  private LoggerRepository configureHierarchy(String logAppenderName)
+      throws IOException {
+    String inetAddressStr = mSocket.getInetAddress().getHostAddress();
+    Properties properties = new Properties();
+    File configFile;
+    try {
+      configFile = new File(new URI(System.getProperty("log4j.configuration")));
+    } catch (URISyntaxException e) {
+      // Alluxio log server cannot derive a valid path to log4j.properties. Since this
+      // properties file is global, we should throw an exception.
+      throw new RuntimeException("Cannot derive a valid URI to log4j.properties file.", e);
+    }
+    try (FileInputStream inputStream = new FileInputStream(configFile)) {
+      properties.load(inputStream);
+    }
+    Level level = Level.INFO;
```

Comment:
does this mean we'd ignore `DEBUG` logs sent by clients?

Reply:
Done.